File: code/setup_AB4.py
import numpy as np
import ipywidgets as widgets
from ipywidgets import interact
import matplotlib.pyplot as plt

import template.output as output
import template.questions as questions
import template.text_order_widget as text_order_widget
import template.table_widget as table


def als_with_mask(matrix, num_factors=2, reg_param=0, max_iter=100):
    """
    Alternating Least Squares (ALS) algorithm for matrix factorization with a mask.

    Parameters:
    - matrix: numpy array
        The target matrix to be factorized.
    - num_factors: int, optional (default=2)
        Number of latent factors.
    - reg_param: float, optional (default=0.01)
        Regularization parameter.
    - max_iter: int, optional (default=100)
        Maximum number of iterations.

    Returns:
    - U: numpy array
        User factors matrix.
    - M: numpy array
        Item factors matrix.
    """

    # Create a mask indicating observed (non-zero) values
    mask = np.where(matrix != 0, 1, 0)

    # Get the shape of the matrix
    num_users, num_items = matrix.shape

    # Initialize user and item factors matrices with random values
    U = np.random.rand(num_users, num_factors)
    M = np.random.rand(num_factors, num_items)

    # ALS algorithm iterations
    error_list = []
    for x in range(max_iter):
        # Update user factors U
        for i in range(num_users):
            observed_indices = np.where(mask[i] == 1)[0]
            M_T_M = np.dot(M[:, observed_indices], M[:, observed_indices].T) + reg_param * np.eye(num_factors)
            M_T_R = np.dot(M[:, observed_indices], matrix[i, observed_indices].T)
            U[i] = np.linalg.solve(M_T_M, M_T_R)

        # Update item factors M
        for j in range(num_items):
            observed_indices = np.where(mask[:, j] == 1)[0]
            U_T_U = np.dot(U[observed_indices].T, U[observed_indices]) + reg_param * np.eye(num_factors)
            U_T_R = np.dot(U[observed_indices].T, matrix[observed_indices, j])
            M[:, j] = np.linalg.solve(U_T_U, U_T_R)

        # Reconstruct the target matrix
        predicted_matrix = np.dot(U, M)
        error_value = calcError(matrix, predicted_matrix)
        error_list.append(error_value)
    return U, M, predicted_matrix, error_list


def gradient_descent_with_mask_(matrix, num_factors=2, reg_param=0, learning_rate=0.01, max_iter=100):
    """
    Gradient Descent algorithm for matrix factorization with a mask.

    Parameters:
    - matrix: numpy array
        The target matrix to be factorized.
    - num_factors: int, optional (default=2)
        Number of latent factors.
    - reg_param: float, optional (default=0.01)
        Regularization parameter.
    - learning_rate: float, optional (default=0.01)
        Learning rate for gradient descent.
    - max_iter: int, optional (default=100)
        Maximum number of iterations.

    Returns:
    - U: numpy array
        User factors matrix.
    - M: numpy array
        Item factors matrix.
    """

    # Create a mask indicating observed (non-zero) values
    mask = np.where(matrix != 0, 1, 0)

    # Get the shape of the matrix
    num_users, num_items = matrix.shape

    # Initialize user and item factors matrices with random values
    U = np.random.rand(num_users, num_factors)
    M = np.random.rand(num_factors, num_items)

    # Gradient Descent algorithm iterations
    error_list = []
    for x in range(max_iter):
        # Update user factors U
        for i in range(num_users):
            observed_indices = np.where(mask[i] == 1)[0]
            predicted_ratings = np.dot(U[i, :], M[:, observed_indices])
            error = matrix[i, observed_indices] - predicted_ratings
            gradient_U = -2 * np.dot(error, M[:, observed_indices].T) + 2 * reg_param * U[i, :]
            U[i, :] -= learning_rate * gradient_U

        # Update item factors M
        for j in range(num_items):
            observed_indices = np.where(mask[:, j] == 1)[0]
            predicted_ratings = np.dot(U[observed_indices, :], M[:, j])
            error = matrix[observed_indices, j] - predicted_ratings
            gradient_M = -2 * np.dot(U[observed_indices, :].T, error) + 2 * reg_param * M[:, j]
            M[:, j] -= learning_rate * gradient_M

        # Reconstruct the target matrix
        predicted_matrix = np.dot(U, M)
        error_value = calcError(matrix, predicted_matrix)
        error_list.append(error_value)

    return U, M, predicted_matrix, error_list


def calcError(matrix, predicted_matrix):
    error = 0
    for i in range(matrix.shape[0]):
        for j in range(matrix.shape[1]):
            if matrix[i, j] != 0:
                error = np.abs(matrix[i][j] - predicted_matrix[i][j])**2 / np.prod(matrix.shape)
    return error


def print_rounded_matrix(matrix):
    for row in matrix:
        for element in row:
            print(round(element, 2), end=" ")
        print()


def print_results(R1, U, M, P, error):
    # Print results
    print("Originale Matrix:")
    print_rounded_matrix(R1)
    print("\nVorhersagen Matrix:")
    print_rounded_matrix(P)
    print("\nMovie Matrix:")
    print_rounded_matrix(M)
    print("\nUser Matrix:")
    print_rounded_matrix(U)
    print("\nFehler auf bekannten Daten:")
    print(error[len(error) - 1])


def plotMSEError(error_list):
    # Erstelle einen Plot
    plt.plot(range(1, len(error_list) + 1), error_list, marker='o', linestyle='-')

    # Beschriftungen für die Achsen und den Plot
    plt.title('Error Plot')
    plt.xlabel('Iteration')
    plt.ylabel('Error Value')

    # Zeige den Plot an
    plt.show()


def checkErrorU(error_u_cal):
    errorU = error_u_cal(1, 1)
    if (errorU is None):
        output.wrong("Setze eine Formel ein")
        return
    u1 = np.array([2, -2, -3, 3, 0, 10])
    u2 = np.array([1, 2, -3, 2, 0.1, -0.4])

    eps = 0.01

    def error_u_sol(u11, u12):
        return (2 - (u11 + u12))**2 + (3 - (u11 + u12))**2

    cal = np.array([error_u_cal(u11, u12) for u11, u12 in zip(u1, u2)])
    sol = np.array([error_u_sol(u11, u12) for u11, u12 in zip(u1, u2)])

    if np.max(np.abs(cal - sol)) > eps:
        output.wrong("Deine Lösung für die Fehlerfunktion ist noch nicht korrekt.\n")
    else:
        output.success("Deine Lösung für die Fehlerfunktion ist korrekt.\n")
        plot_error(error_u_cal, -1, 3)


def check_gradient_u(error_gradient1, error_gradient2):

    u1 = np.array([2, -2, -3, 3, 0, 10])
    u2 = np.array([1, 2, -3, 2, 0.1, -0.4])

    eps = 0.01

    def error_gradient1_sol(u11, u12):
        return -2 * (2 - u11 - u12) - 2 * (3 - u11 - u12)

    def error_gradient2_sol(u11, u12):
        return -2 * (2 - u11 - u12) - 2 * (3 - u11 - u12)

    # u11
    if (error_gradient1 is None):
        output.wrong("Setze eine Formel ein")
        return
    cal1 = np.array([error_gradient1(u11, u12) for u11, u12 in zip(u1, u2)])
    sol1 = np.array([error_gradient1_sol(u11, u12) for u11, u12 in zip(u1, u2)])

    if np.max(np.abs(cal1 - sol1)) > eps:
        output.wrong("Deine Lösung für die Ableitung nach u11 ist nicht noch nicht korrekt.")
    else:
        output.success("Deine Lösung für die Ableitung nach u11 ist korrekt.")

    # u12
    if (error_gradient2 is None):
        output.wrong("Setze eine Formel ein")
        return
    cal2 = np.array([error_gradient2(u11, u12) for u11, u12 in zip(u1, u2)])
    sol2 = np.array([error_gradient2_sol(u11, u12) for u11, u12 in zip(u1, u2)])

    if np.max(np.abs(cal2 - sol2)) > eps:
        output.wrong("Deine Lösung für die Ableitung nach u12 ist nicht noch nicht korrekt.")
    else:
        output.success("Deine Lösung für die Ableitung nach u12 ist korrekt.")


def check_min_u(u11_min, u12_min, error_u, error_gradient1, error_gradient2):

    eps = 0.0001

    if abs(error_gradient1(u11_min, u12_min)) < eps and abs(error_gradient2(u11_min, u12_min)) < eps:
        output.success("Dein Ergebnis ist korrekt.")
        
        # Evaluate the error function at the minimum
        error = error_u(u11_min, u12_min)
        print(f"Einsetzen der gefundenen Lösung u11 = {u11_min} und u12 = {u12_min} in die Fehlerfunktion liefert einen Fehler von {error}.")
    else:
        output.wrong("Deine gefundenen Einträge von U liefern noch nicht das Minimum der Fehlerfunktion.")




def check_min_m(m11_min, m12_min, m21_min, m22_min, error_m, error_gradient_m11, error_gradient_m12, error_gradient_m21, error_gradient_m22):

    eps = 0.0001

    if (abs(error_gradient_m11(m11_min, m12_min, m21_min, m22_min)) < eps and 
        abs(error_gradient_m12(m11_min, m12_min, m21_min, m22_min)) < eps and 
        abs(error_gradient_m21(m11_min, m12_min, m21_min, m22_min)) < eps and 
        abs(error_gradient_m22(m11_min, m12_min, m21_min, m22_min)) < eps):
        
        output.success("Dein Ergebnis ist korrekt.")
        
        # Evaluate the error function at the minimum
        error = error_m(m11_min, m12_min, m21_min, m22_min)
        print(f"Einsetzen der gefundenen Lösung m11 = {m11_min}, m12 = {m12_min}, m21 = {m21_min} und m22 = {m22_min} in die Fehlerfunktion liefert einen Fehler von {error}.")
    else:
        output.wrong("Deine gefundenen Werte für die Einträge von U liefern noch nicht das Minimum der Fehlerfunktion.")

        
def check_gradient_m(error_gradient_m11, error_gradient_m12, error_gradient_m21, error_gradient_m22, u1min, u2min):

    m1 = np.array([2, -2, -3, 3, 0, 10])
    m2 = np.array([1, 2, -3, 2, 0.1, -0.4])
    m3 = np.array([0, 2, -3, 2, 0.1, -0.4])
    m4 = np.array([-10, 2, -3, 2, 2.1, 0.4])
    
    eps = 0.01

    def error_gradient_m11_sol(m11, m12, m21, m22):
        return 2 * (2 - u1min * m11 - u2min * m21) * (-u1min)

    def error_gradient_m12_sol(m11, m12, m21, m22):
        return 2 * (3 - u1min * m12 - u2min * m22) * (-u1min)

    def error_gradient_m21_sol(m11, m12, m21, m22):
        return 2 * (2 - u1min * m11 - u2min * m21) * (-u2min)

    def error_gradient_m22_sol(m11, m12, m21, m22):
        return 2 * (3 - u1min * m12 - u2min * m22) * (-u2min)

    # m11
    if np.isnan(error_gradient_m11(1, 2, 3, 4)):
        output.wrong("Gib die partielle Ableitung nach m11 für das erste NaN ein.")
    else:
        cal1 = np.array([error_gradient_m11(m1[i], m2[i], m3[i], m4[i]) for i in range(len(m1))])
        sol1 = np.array([error_gradient_m11_sol(m1[i], m2[i], m3[i], m4[i]) for i in range(len(m1))])
        
        if np.max(np.abs(cal1 - sol1)) > eps:
            output.wrong("Deine Lösung für die Ableitung nach m11 ist nicht noch nicht korrekt.")
        else:
            output.success("Deine Lösung für die Ableitung nach m11 ist korrekt.")

    # m12
    if np.isnan(error_gradient_m12(1, 2, 3, 4)):
        output.wrong("Gib die partielle Ableitung nach m12 für das zweite NaN ein.")
    else:
        cal2 = np.array([error_gradient_m12(m1[i], m2[i], m3[i], m4[i]) for i in range(len(m1))])
        sol2 = np.array([error_gradient_m12_sol(m1[i], m2[i], m3[i], m4[i]) for i in range(len(m1))])
        
        if np.max(np.abs(cal2 - sol2)) > eps:
            output.wrong("Deine Lösung für die Ableitung nach m12 ist nicht noch nicht korrekt.")
        else:
            output.success("Deine Lösung für die Ableitung nach m12 ist korrekt.")

    # m21
    if np.isnan(error_gradient_m21(1, 2, 3, 4)):
        output.wrong("Gib die partielle Ableitung nach m21 für das dritte NaN ein.")
    else:
        cal3 = np.array([error_gradient_m21(m1[i], m2[i], m3[i], m4[i]) for i in range(len(m1))])
        sol3 = np.array([error_gradient_m21_sol(m1[i], m2[i], m3[i], m4[i]) for i in range(len(m1))])
        
        if np.max(np.abs(cal3 - sol3)) > eps:
            output.wrong("Deine Lösung für die Ableitung nach m21 ist nicht noch nicht korrekt.")
        else:
            output.success("Deine Lösung für die Ableitung nach m21 ist korrekt.")

    # m22
    if np.isnan(error_gradient_m22(1, 2, 3, 4)):
        output.wrong("Gib die partielle Ableitung nach m22 für das vierte NaN ein.")
    else:
        cal4 = np.array([error_gradient_m22(m1[i], m2[i], m3[i], m4[i]) for i in range(len(m1))])
        sol4 = np.array([error_gradient_m22_sol(m1[i], m2[i], m3[i], m4[i]) for i in range(len(m1))])
        
        if np.max(np.abs(cal4 - sol4)) > eps:
            output.wrong("Deine Lösung für die Ableitung nach m22 ist nicht noch nicht korrekt.")
        else:
            output.success("Deine Lösung für die Ableitung nach m22 ist korrekt.")

def print_UTimesM(u11_min, u12_min, m11_min, m12_min, m21_min, m22_min):
    U = np.array([[u11_min, u12_min]])
    M = np.array([[m11_min, m12_min], [m21_min, m22_min]])
    
    print("\nDeine Zerlegung in U und M lautet:")
    print("\nUsermatrix U:")
    print(U)
    
    print("\nMovie-Matrix M:")
    print(M)
    
    P = np.dot(U, M)
    
    print("\nDas Produkt U*M und damit die Vorhersagematrix P lautet:")
    print(P)

def plot_error(error_u_cal, start, end):
    import matplotlib.pyplot as plt
    from mpl_toolkits.mplot3d import Axes3D

    u11 = np.linspace(start, end, 100)
    u12 = np.linspace(start, end, 100)
    U11, U12 = np.meshgrid(u11, u12)
    Z = np.array([error_u_cal(u11, u12) for u11, u12 in zip(np.ravel(U11), np.ravel(U12))])
    Z = Z.reshape(U11.shape)

    fig = plt.figure()
    ax = fig.add_subplot(111, projection='3d')
    ax.plot_surface(U11, U12, Z, cmap='viridis')
    ax.set_xlabel('u11')
    ax.set_ylabel('u12')
    ax.set_zlabel('Error')
    plt.show()


def promt_4_2_a1():
    questions.prompt_answer("WS4-2a1", input_prompt="Wert für $\lambda$", input_description="Deine Antwort")


def promt_4_2_a2():
    questions.prompt_answer("WS4-2a2", input_prompt="Werte für U und M", input_description="Deine Antwort")

File: code/setup_AB4.jl

using Suppressor, DelimitedFiles 

@suppress begin

using LaTeXStrings
using LinearAlgebra
using Calculus
using StatsBase
using YAML
using Random
using PrettyTables

# using NPZ, SparseArrays, NMF, PrettyTables, DataFrames

include("output.jl")
include("utils.jl")
include("utils_ALS.jl")
include("utils_plots.jl")
include("utils_pretty_tables.jl");
include("utils_plotlyJS.jl");
include("utils_split_data.jl");
include("loss_functions.jl");
include("utils_lsqFIT.jl")
include("setup_language.jl")

end


# Definition Ratingmatrix Aufgabe 2
# Bei dieser Matrix steigt error wieder:
RS = [3 1 0 3 1 0;
     1 0 4 1 3 1;
     0 1 1 0 1 4;
     0 0 5 0 4 0]

# AB 5:
ROLD = [3 1 0 4 1 0 2 4 1 ;
     1 0 4 1 3 1 0 0 5;
     0 1 1 0 1 4 2 0 5;
     1 0 3 0 2 0 1 0 5;
     3 1 0 4 1 0 1 4 1;
     0 0 5 0 4 0 0 0 5 ]



# Überprüfefunktionen 

 function computeErrorSol(R,U,M)  
    
    nonZeros = findall(!isnan, R)
    sum((((R - U * M).^2))[nonZeros])/length(R)
end


function checkErrorU(errorUCal)
    
    if isnan(errorUCal(1,1))
        printError("Gib deine Formel für NaN ein.")
        
    else
    
        u1 = [2,-2,-3, 3,0,10]
        u2 = [1, 2,-3, 2, 0.1, -0.4]

        eps = 0.01;

        errorUSol(u11,u12) = (2 - (u11 + u12))^2 + (3 - (u11 + u12))^2;

        cal = errorUCal.(u1,u2)
        sol = errorUSol.(u1,u2)

        if maximum(abs.(cal - sol)) > eps 

            printError("Deine Lösung für die Fehlerfunktion ist noch nicht noch nicht korrekt.\n")
        else
            printSuccess("Deine Lösung für die Fehlerfunktion ist korrekt.\n")
            plotError(errorUCal,-1,3)
        end
    end
end


function checkGradientU(ErrorGradient1,ErrorGradient2)
    u1 = [2,-2,-3, 3,0,10]
    u2 = [1, 2,-3, 2, 0.1, -0.4]
    
    eps = 0.01;
    
    ErrorGradient1Sol(u11,u12) =  -2*(2-u11-u12)-2*(3-u11-u12);
    ErrorGradient2Sol(u11,u12) =  -2*(2-u11-u12)-2 *(3-u11-u12);
    
    
    # u11
    if isnan(ErrorGradient1(1,2))
        printError("Gib die partielle Ableitung nach u11 für das erste NaN ein.")
        
    else
        cal1 = ErrorGradient1.(u1,u2)
        sol1 = ErrorGradient1Sol.(u1,u2)
        
         if maximum(abs.(cal1 - sol1)) > eps 
        
            printError("Deine Lösung für die Ableitung nach u11 ist nicht noch nicht korrekt.")
        else
            printSuccess("Deine Lösung für die Ableitung nach u11 ist korrekt.")
        end
    end
    
    
    # u12
    if isnan(ErrorGradient2(1,2))
        printError("Gib die partielle Ableitung nach u12 für das zweite NaN ein.")
        
    else
        cal2 = ErrorGradient2.(u1,u2)
        sol2 = ErrorGradient2Sol.(u1,u2)
    
        
        if maximum(abs.(cal2 - sol2)) > eps 

            printError("Deine Lösung für die Ableitung nach u12 ist nicht noch nicht korrekt.")
        else
            printSuccess("Deine Lösung für die Ableitung nach u12 ist korrekt.")
        end
    end
    
end


function checkMinU(u11_min,u12_min,errorU)
    
    eps = 0.0001;
    
    # Ableitung bezüglich u11:
    ErrorGradient1(u11,u12) = +2*(2-u11-u12)*(-1) + 2*(3-u11-u12)* (-1); # Ableitung nach u11

    # Ableitung bezüglich u12:
    ErrorGradient2(u11,u12) =  +2*(2-u11-u12)*(-1) + 2 *(3-u11-u12)*(-1) ; # Ableitung nach u12
    
    
    if abs(ErrorGradient1(u11_min, u12_min)) < eps && abs(ErrorGradient2(u11_min, u12_min)) < eps

        printSuccess("Dein Ergebnis ist korrekt.")
        
        # Auswertung der Fehlerfunktion am Minimum
        error = errorU(u11_min, u12_min);
        println("Einsetzen der gefundenen Lösung u11 = $u11_min und u12 = $u12_min in die Fehlerfunktion liefert einen Fehler von $error."); 
        
    else
        printError("Deine gefundenen Einträge von U liefern noch nicht das Minimum der Fehlerfunktion.")
    end
    
end



function checkMinU(u11_min,u12_min,errorU, ErrorGradient1, ErrorGradient2)
    
    eps = 0.0001;

    if abs(ErrorGradient1(u11_min, u12_min)) < eps && abs(ErrorGradient2(u11_min, u12_min)) < eps

        printSuccess("Dein Ergebnis ist korrekt.")
        
        # Auswertung der Fehlerfunktion am Minimum
        error = errorU(u11_min, u12_min);
        println("Einsetzen der gefundenen Lösung u11 = $u11_min und u12 = $u12_min in die Fehlerfunktion liefert einen Fehler von $error."); 
        
    else
        printError("Deine gefundenen Einträge von U liefern noch nicht das Minimum der Fehlerfunktion.")
    end
    
end



function checkGradientM(ErrorGradientM11,ErrorGradientM12,ErrorGradientM21,ErrorGradientM22,u1min,u2min)
    
    m1 = [2,-2,-3, 3,0,10]
    m2 = [1, 2,-3, 2, 0.1, -0.4]
    m3 = [0, 2,-3, 2, 0.1, -0.4]
    m4 = [-10, 2,-3, 2, 2.1, 0.4]
    
    eps = 0.01;

    ErrorGradientM11Sol(m11,m12,m21,m22) =  2 * (2 - u1min * m11 - u2min* m21)*(-u1min);
    ErrorGradientM12Sol(m11,m12,m21,m22) =  2 * (3 - u1min * m12 - u2min * m22)*(-u1min);
    ErrorGradientM21Sol(m11,m12,m21,m22) =  2 * (2 - u1min * m11 - u2min* m21)*(-u2min);
    ErrorGradientM22Sol(m11,m12,m21,m22) =  2 * (3 - u1min * m12 - u2min * m22)*(-u2min);
    
    
    # m11
    if isnan(ErrorGradientM11(1,2,3,4))
        printError("Gib die partielle Ableitung nach m11 für das erste NaN ein.")
        
    else
        cal1 = ErrorGradientM11.(m1,m2,m3,m4)
        sol1 = ErrorGradientM11Sol.(m1,m2,m3,m4)
        
        if maximum(abs.(cal1 - sol1)) > eps 

            printError("Deine Lösung für die Ableitung nach m11 ist nicht noch nicht korrekt.")
        else
            printSuccess("Deine Lösung für die Ableitung nach m11 ist korrekt.")
        end  
    end
    
    
    # m12
    if isnan(ErrorGradientM12(1,2,3,4))
        printError("Gib die partielle Ableitung nach m12 für das zweite NaN ein.")
        
    else
        cal2 = ErrorGradientM12.(m1,m2,m3,m4)
        sol2 = ErrorGradientM12Sol.(m1,m2,m3,m4)
        
        if maximum(abs.(cal2 - sol2)) > eps 
        
            printError("Deine Lösung für die Ableitung nach m12 ist nicht noch nicht korrekt.")
        else
            printSuccess("Deine Lösung für die Ableitung nach m12 ist korrekt.")
        end 
    end
        
    
    # m21
    if isnan(ErrorGradientM21(1,2,3,4))
        printError("Gib die partielle Ableitung nach m21 für das dritte NaN ein.")
        
    else
        cal3 = ErrorGradientM21.(m1,m2,m3,m4)
        sol3 = ErrorGradientM21Sol.(m1,m2,m3,m4)
        
        if maximum(abs.(cal3 - sol3)) > eps 

            printError("Deine Lösung für die Ableitung nach m21 ist nicht noch nicht korrekt.")
        else
            printSuccess("Deine Lösung für die Ableitung nach m21 ist korrekt.")
        end
    end

        
    # m22
    if isnan(ErrorGradientM22(1,2,3,4))
        printError("Gib die partielle Ableitung nach m22 für das vierte NaN ein.")
        
    else
        cal4 = ErrorGradientM22.(m1,m2,m3,m4)
        sol4 = ErrorGradientM22Sol.(m1,m2,m3,m4)
        
        if maximum(abs.(cal4 - sol4)) > eps 

            printError("Deine Lösung für die Ableitung nach m22 ist nicht noch nicht korrekt.")
        else
            printSuccess("Deine Lösung für die Ableitung nach m22 ist korrekt.")
        end
    end
    
end


function checkMinM(m11_min,m12_min,m21_min, m22_min, errorM, ErrorGradientM11, ErrorGradientM12, ErrorGradientM21, ErrorGradientM22)
    
    eps = 0.0001;

    if abs(ErrorGradientM11(m11_min, m12_min,m21_min,m22_min)) < eps && abs(ErrorGradientM12(m11_min, m12_min,m21_min,m22_min)) < eps && abs(ErrorGradientM21(m11_min, m12_min,m21_min,m22_min)) < eps && abs(ErrorGradientM22(m11_min, m12_min,m21_min,m22_min)) < eps 

        printSuccess("Dein Ergebnis ist korrekt.")
        
        # Auswertung der Fehlerfunktion am Minimum
        error = errorM(m11_min, m12_min, m21_min, m22_min);
        println("Einsetzen der gefundenen Lösung m11 = $m11_min, m12 = $m12_min, m21 = $m21_min und m22 = $m22_min in die Fehlerfunktion liefert einen Fehler von $error."); 
        
    else
        printError("Deine gefundenen Werte für die Einträge von U liefern noch nicht das Minimum der Fehlerfunktion.")
    end
    
end 

function printSave()
         println("Die  Rating-Matrix R1 wurde gespeichert.")
end


function printUtimesM()

    U = [u11_min u12_min]
    M = [m11_min m12_min; m21_min m22_min]
    println("\nDeine Zerlegung in U und M lautet:")
    println("\nUsermatrix U:")
    printSimple(U)
    println("\nMovie-Matrix M:")
    printSimple(M)
    println("\nDas Produkt U*M und damit die Vorhersagematrix P lautet:")
    P = U*M
    printSimple(P)

end






function catchErrorALS()
    println("\n________________________________________________\n\nFalls du eine Fehlermeldung erhältst, lies die nächste Teilaufgabe auf diesem Arbeitsblatt!\n________________________________________________\n")
end



function computeALS(n_iters, n_factors, train_matrix_default, test_matrix_default, lambda,normalization,datasplit, verbose = false)
    
""" 
    n_iters   = number of iterations to train the algorithm
    n_factors = number of latent factors to use in matrix 
    average = average of the ratings in the training matrix
    lambda    = regularization term for item/user latent factors 
    normalization = binary variable. true = training data is normalized, false otherwise
    datasplit = binary variable. true = data was split into test and train data, false = factorization on the whole data 
"""
   
    test_mse_record  = [] 
    train_mse_record = [] 
    PredictionMatrix = []
    
    
    if isnan(train_matrix_default[1]) 
        printError("Wähle eine Rating-Matrix.")
        
    elseif isnan(n_iters) || (n_iters % 1) != 0 || (n_iters)<0
        printError("Gib eine natürliche Zahl für die Variable iterations ein.") 
        
    elseif isnan(n_factors)
        printError("Gib eine natürliche Zahl für die Variable features ein.")
        
    elseif isnan(lambda)
        printError("Gib eine Zahl größer 0 für die Variable lambda ein.")    
        
    else
    
    
    train_matrix_default = convert(Array{Float64,2},train_matrix_default)
    idx = findall(!iszero, train_matrix_default)
    D = zeros(size(train_matrix_default));
    D[idx] .= 1  
        
        
        if normalization == false    
            average = 0
            train_matrix = train_matrix_default[:,:]

        else 

            idx = findall(!iszero, train_matrix_default)
            average = mean(train_matrix_default[idx])
            train_matrix = train_matrix_default[:,:];
            train_matrix[idx] = train_matrix[idx] .- average
        end

        
        Random.seed!(123)
        U = rand(size(train_matrix, 1), n_factors) 
        M = rand(n_factors, size(train_matrix, 2))  
 
        for k = 1 : n_iters

            if k % 2 == 0 
                M = als_step_m_nonzeros_LSQFIT(M, U, train_matrix, lambda,D)[:,:];  
                if verbose println("\nWiederholung $k: Die User-Matrix ist fest. Die Movie-Matrix wird optimiert.") end
    
            else 
                U = als_step_u_nonzeros_LSQFIT(U, M, train_matrix, lambda,D)[:,:];
                if verbose println("\nWiederholung $k: Die Movie-Matrix ist fest. Die User-Matrix wird optimiert.") end
            end
            
            predictions = predict(U , M).+average
            train_mse = compute_mse(train_matrix, predictions); 
            append!(train_mse_record, train_mse); 
        
            if datasplit == true
                test_matrix = convert(Array{Float64,2},test_matrix_default)
                test_mse = compute_mse(test_matrix, predictions);
                append!(test_mse_record, test_mse);
            else
            
            end

        end
        
        PredictionMatrix = U * M
        PredictionMatrix = round.(PredictionMatrix, digits = 4)

        if datasplit == false
        println("\n________________________________________________\nGemittelte Summe der Fehlerquadrate nach $iterations Wiederholungen: \n$(round(train_mse_record[n_iters], digits =5))\n")


        else
            println("\n________________________________________________\nGemittelte Summe der Fehlerquadrate auf den Trainingsdaten nach $iterations Wiederholungen: \n$(round(train_mse_record[n_iters], digits =5))\n")
     #        println("\nGemittelte Summe der Fehlerquadrate auf den Testdaten nach $iterations Wiederholungen: \n$(round(test_mse_record[n_iters], digits =5))\n")  

        end

       if verbose println("Hinweis: Um den Fehler auf verschieden großen Rating-Matrizen besser vergleichen zu können, wurde die Summe der Fehlerquadrate  gemittelt. Dazu wurde der Fehler durch die Anzahl der bekannten Ratings geteilt. Wir betrachten also die gemittelte Summe der Fehlerquadrate.\n") end
    
    end
    
    return train_mse_record, PredictionMatrix, U, M
    
end


